Connect to Amazon Polly (#7248)
aws-amplify/docs: src/pages/[platform]/build-a-backend/data/custom-business-logic/connect-amazon-polly/index.mdx

Reviewer: Since this is a new file, don't we need to add it to `directory.mjs`?

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {

>   title: 'Connect to Amazon Polly',

Reviewer 2: ```suggestion
  title: 'Connect to Amazon Polly for Text-To-Speech APIs',
```

  description:
    'Connect to Amazon Polly.',
  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps() {
  return {
    props: {
      meta
    }
  };
}

> # Connect to Amazon Polly

Reviewer 2: ```suggestion
# Connect to Amazon Polly for Text-To-Speech APIs
```

Key features of Amazon Polly include:

- **Multiple Voices and Languages**: Amazon Polly supports dozens of voices across various languages and dialects, giving you the flexibility to choose the most appropriate voice for your use case.

- **High-Quality Speech**: Amazon Polly's neural and standard voices offer natural and realistic speech quality.

- **Speech Marks and SSML**: Customize your speech output with Speech Synthesis Markup Language (SSML) tags and obtain speech timing information with speech marks.

- **Scalable and Cost-Effective**: Amazon Polly's pay-as-you-go pricing model makes it a cost-effective solution for adding speech capabilities to your applications.

In this section, you'll learn how to integrate Amazon Polly into your application using AWS Amplify, enabling you to leverage its powerful text-to-speech capabilities seamlessly.

## Step 1 - Setup the project

> Set up your project by following the instructions in the (Quickstart guide)[https://docs.amplify.aws/javascript/start/quickstart/]. 

Reviewer: The link is formatted incorrectly. The square braces should be around the link text.

Also, did you mean to link directly to `javascript` or should this link be platform agnostic?

From the look of the platforms, it should be platform agnostic so we could do this instead:
```suggestion
Set up your project by following the instructions in the [Quickstart guide](/[platform]/start/quickstart/). 
```

## Step 2 - Install Polly Libraries

>  To install the Amazon Polly library, run the following command in your project's root folder:

Reviewer 2: ```suggestion
We'll create a new API endpoint that'll use the the AWS SDK to call the Amazon Polly service. To install the Amazon Polly SDK, run the following command in your project's root folder:
```

    .arguments({
      text: a.string().required()
    })
    .returns(a.string().required())
    .authorization([a.allow.public()])
    .handler(a.handler.function(convertTextToSpeech))
});

export type Schema = ClientSchema<typeof schema>;

export const data = defineData({
  schema,
  authorizationModes: {
    defaultAuthorizationMode: 'apiKey',
    // API Key is used for a.allow.public() rules
    apiKeyAuthorizationMode: {
      expiresInDays: 30,
    },
  },
});

```
## Step 7 - Update Storage permissions

Customize your storage settings to manage access to various paths within your storage bucket. It's necessary to update the Storage resource to provide access to the `convertTextToSpeech` resource. Modify the file `amplify/storage/resource.ts` as shown below.

```ts title="amplify/storage/resource.ts"
import { defineStorage } from "@aws-amplify/backend"
import { convertTextToSpeech } from "../data/resource"

export const storage = defineStorage({
  name: "predictions_gen2",
  access: allow => ({
    '/public/*': [
      allow.resource(convertTextToSpeech).to(['write']),
      allow.guest.to(['read', 'write', 'get'])
    ]
  })
})
```

## Step 8 - Configure the frontend

> Import and load the configuration file in your app. It's recommended you add the Amplify configuration step to your app's root entry point. 

Reviewer 2: this text seems to be copied/pasted into the wrong place. The code below doesn't show the configuration file.

Author: added the config code

- **High-Quality Speech**: Amazon Polly's neural and standard voices offer natural and realistic speech quality.

- **Speech Marks and Speech Synthesis Markup Language**: Customize your speech output with Speech Synthesis Markup Language tags and obtain speech timing information with speech marks.

- **Scalable and Cost-Effective**: Amazon Polly's pay-as-you-go pricing model makes it a cost-effective solution for adding speech capabilities to your applications.

In this section, you'll learn how to integrate Amazon Polly into your application using AWS Amplify, enabling you to leverage its powerful text-to-speech capabilities seamlessly.

## Step 1 - Setup the project

Set up your project by following the instructions in the [Quickstart guide](/[platform]/start/quickstart/). 

## Step 2 - Install Polly Libraries
We'll create a new API endpoint that'll use the the AWS SDK to call the Amazon Polly service. To install the Amazon Polly SDK, run the following command in your project's root folder:

```bash
npm install @aws-sdk/client-polly

> 

Reviewer 2: ```suggestion
```

  actions: ['polly:StartSpeechSynthesisTask'],
  resources: ['*']
}))

```
## Step 5 - Configure the function handler 

Define the function handler by creating a new file, `amplify/data/convertTextToSpeech.ts`. This function converts text into speech using Amazon Polly and stores the synthesized speech as an MP3 file in an S3 bucket.

```ts title="amplify/data/convertTextToSpeech.ts"

import { Schema } from "./resource";
import { PollyClient, StartSpeechSynthesisTaskCommand } from '@aws-sdk/client-polly'
import { env } from '$amplify/env/convertTextToSpeech'

export const handler: Schema["convertTextToSpeech"]["functionHandler"] = async (event) => {
  const client = new PollyClient()
  const task = new StartSpeechSynthesisTaskCommand({
    OutputFormat: 'mp3',
    SampleRate: '8000',
    Text: event.arguments.text,
    TextType: 'text',
    VoiceId: 'Amy',
    OutputS3BucketName: env.predictions_gen2_BUCKET_NAME,
    OutputS3KeyPrefix: 'public/'
  })
  const result = await client.send(task)

  return result.SynthesisTask?.OutputUri?.replace(
    'https://s3.us-east-1.amazonaws.com/' + env.predictions_gen2_BUCKET_NAME + '/public/', "") ?? ""
}

```

> ## Step 6 - Define the Custom Mutation and function

Reviewer 2: ```suggestion
## Step 6 - Define the custom mutation and function
```

In your `amplify/data/resource.ts` file, define the function using defineFunction and then reference the function with your mutation using a.handler.function() as a handler.

```ts title="amplify/data/resource.ts"
import { type ClientSchema, a, defineData, defineFunction } from '@aws-amplify/backend';

export const convertTextToSpeech = defineFunction({
  entry: './convertTextToSpeech.ts'
})

const schema = a.schema({
  convertTextToSpeech: a.mutation()
    .arguments({
      text: a.string().required()
    })
    .returns(a.string().required())

>     .authorization([a.allow.public()])

Reviewer 2: ```suggestion
    .authorization(allow => [allow.publicApiKey()])
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Connect to Amazon Polly for Text-To-Speech APIs',
  description:
    'Connect to Amazon Polly.',
  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps() {
  return {
    props: {
      meta
    }
  };
}

> # Connect to Amazon Polly for Text-To-Speech APIs

Reviewer: We can remove this title since the docs repo will automatically add the `meta.title` to the top of the page. This is what the mdx looks like at the moment, there are two of the same headings:

![image](https://github.com/aws-amplify/docs/assets/54393192/b253b8e7-2ff9-44a1-a3c2-e9fef9c920de)

Author: removed
